File: sdks/python/docflow/rag/chunker.py
"""RAG chunker for DocFlow."""

import re
from typing import List, Optional

from ..types import Chunk, ChunkMetadata, RAGConfig


class RAGChunker:
    """Smart chunker for RAG-optimized markdown content.
    
    Features:
    - Respects heading boundaries
    - Keeps tables together
    - Maintains image+description pairs
    - Adds overlap for context
    
    Example:
        >>> chunker = RAGChunker(config=RAGConfig())
        >>> chunks = chunker.chunk(markdown_content)
    """

    def __init__(self, config: Optional[RAGConfig] = None) -> None:
        """Initialize chunker.
        
        Args:
            config: RAG configuration options.
        """
        self.config = config or RAGConfig()

    def chunk(self, markdown: str) -> List[Chunk]:
        """Split markdown into RAG-optimized chunks.
        
        Args:
            markdown: Markdown content to chunk.
            
        Returns:
            List of Chunk objects.
        """
        # Remove frontmatter for chunking (will be added to metadata)
        content, frontmatter = self._extract_frontmatter(markdown)
        
        if self.config.respect_headings:
            sections = self._split_by_headings(content)
        else:
            sections = [("", content)]
        
        chunks = []
        chunk_index = 0
        char_offset = 0
        
        for section_title, section_content in sections:
            section_chunks = self._chunk_section(
                section_content,
                section_title,
                chunk_index,
                char_offset,
            )
            chunks.extend(section_chunks)
            chunk_index += len(section_chunks)
            char_offset += len(section_content)
        
        # Add overlap between chunks
        if self.config.chunk_overlap > 0:
            chunks = self._add_overlap(chunks, content)
        
        # Add chunk markers if requested
        if self.config.add_chunk_markers:
            for i, chunk in enumerate(chunks):
                chunk.content = f"{chunk.content}\n\n<!-- chunk_boundary: {i} -->"
        
        return chunks

    def _extract_frontmatter(self, markdown: str) -> tuple:
        """Extract YAML frontmatter from markdown."""
        if markdown.startswith("---"):
            end = markdown.find("---", 3)
            if end > 0:
                frontmatter = markdown[3:end].strip()
                content = markdown[end + 3:].strip()
                return content, frontmatter
        return markdown, ""

    def _split_by_headings(self, content: str) -> List[tuple]:
        """Split content by headings."""
        # Pattern for markdown headings
        heading_pattern = re.compile(r'^(#{1,6})\s+(.+)$', re.MULTILINE)
        
        sections = []
        last_end = 0
        last_title = ""
        
        for match in heading_pattern.finditer(content):
            # Content before this heading
            if match.start() > last_end:
                section_content = content[last_end:match.start()].strip()
                if section_content:
                    sections.append((last_title, section_content))
            
            last_title = match.group(2).strip()
            last_end = match.start()
        
        # Remaining content
        if last_end < len(content):
            remaining = content[last_end:].strip()
            if remaining:
                sections.append((last_title, remaining))
        
        if not sections:
            sections = [("", content)]
        
        return sections

    def _chunk_section(
        self,
        content: str,
        section_title: str,
        start_index: int,
        char_offset: int,
    ) -> List[Chunk]:
        """Chunk a section respecting tables and code blocks."""
        chunks = []
        
        # Find special blocks that should stay together
        protected_blocks = self._find_protected_blocks(content)
        
        # Simple chunking by size
        current_chunk = ""
        current_start = char_offset
        
        lines = content.split("\n")
        i = 0
        
        while i < len(lines):
            line = lines[i]
            
            # Check if we're starting a protected block
            block_end = None
            for start, end, block_type in protected_blocks:
                if i == start:
                    block_end = end
                    break
            
            if block_end is not None:
                # Add entire protected block
                block_content = "\n".join(lines[i:block_end + 1])
                
                # If adding this would exceed chunk size, save current and start new
                if len(current_chunk) + len(block_content) > self.config.chunk_size and current_chunk:
                    chunks.append(self._create_chunk(
                        current_chunk.strip(),
                        start_index + len(chunks),
                        current_start,
                        current_start + len(current_chunk),
                        section_title,
                    ))
                    current_chunk = ""
                    current_start = char_offset + sum(len(l) + 1 for l in lines[:i])
                
                current_chunk += block_content + "\n"
                i = block_end + 1
            else:
                # Regular line
                if len(current_chunk) + len(line) > self.config.chunk_size and current_chunk:
                    chunks.append(self._create_chunk(
                        current_chunk.strip(),
                        start_index + len(chunks),
                        current_start,
                        current_start + len(current_chunk),
                        section_title,
                    ))
                    current_chunk = ""
                    current_start = char_offset + sum(len(l) + 1 for l in lines[:i])
                
                current_chunk += line + "\n"
                i += 1
        
        # Don't forget remaining content
        if current_chunk.strip():
            chunks.append(self._create_chunk(
                current_chunk.strip(),
                start_index + len(chunks),
                current_start,
                current_start + len(current_chunk),
                section_title,
            ))
        
        return chunks

    def _find_protected_blocks(self, content: str) -> List[tuple]:
        """Find blocks that should stay together (tables, code blocks)."""
        blocks = []
        lines = content.split("\n")
        
        i = 0
        while i < len(lines):
            line = lines[i].strip()
            
            # Code blocks
            if line.startswith("```"):
                start = i
                i += 1
                while i < len(lines) and not lines[i].strip().startswith("```"):
                    i += 1
                blocks.append((start, i, "code"))
            
            # Tables
            elif line.startswith("|") and line.endswith("|"):
                start = i
                while i < len(lines) and lines[i].strip().startswith("|"):
                    i += 1
                blocks.append((start, i - 1, "table"))
                continue
            
            i += 1
        
        return blocks

    def _create_chunk(
        self,
        content: str,
        index: int,
        start_char: int,
        end_char: int,
        section_title: str,
    ) -> Chunk:
        """Create a Chunk object with metadata."""
        has_table = "|" in content and "---" in content
        has_image = "![" in content or "[Image:" in content
        
        metadata = ChunkMetadata(
            section_title=section_title,
            heading_path=self._extract_heading_path(content),
            has_table=has_table,
            has_image=has_image,
        )
        
        return Chunk(
            content=content,
            index=index,
            start_char=start_char,
            end_char=end_char,
            metadata=metadata,
        )

    def _extract_heading_path(self, content: str) -> List[str]:
        """Extract heading hierarchy from content."""
        headings = []
        for line in content.split("\n"):
            if line.startswith("#"):
                level = len(line.split()[0])
                text = line.lstrip("#").strip()
                headings.append(text)
        return headings

    def _add_overlap(self, chunks: List[Chunk], full_content: str) -> List[Chunk]:
        """Add overlap between chunks for context."""
        if len(chunks) <= 1:
            return chunks
        
        overlap_size = self.config.chunk_overlap
        
        for i in range(1, len(chunks)):
            prev_chunk = chunks[i - 1]
            curr_chunk = chunks[i]
            
            # Get overlap from end of previous chunk
            overlap_text = prev_chunk.content[-overlap_size:]
            
            # Find a good break point (end of sentence or paragraph)
            for break_char in ["\n\n", ". ", "\n"]:
                idx = overlap_text.find(break_char)
                if idx > 0:
                    overlap_text = overlap_text[idx + len(break_char):]
                    break
            
            if overlap_text.strip():
                curr_chunk.content = f"[...] {overlap_text.strip()}\n\n{curr_chunk.content}"
        
        return chunks

File: sdks/go/docflow/rag/chunker.go
package rag

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/xgaslan/docflow/sdks/go/docflow/config"
)

// Chunker splits text into chunks based on configuration.
type Chunker struct {
	Config config.RAGConfig
}

// NewChunker creates a new chunker.
func NewChunker(config config.RAGConfig) *Chunker {
	return &Chunker{
		Config: config,
	}
}

// Chunk represents a text chunk.
// Reusing logic from types.go which defines Chunk struct.

// Chunk splits markdown content into RAG-optimized chunks.
func (c *Chunker) Chunk(markdown string) []Chunk {
	// Extract frontmatter
	content, _ := c.extractFrontmatter(markdown)

	// Split by headings if configured
	var sections []sectionData
	if c.Config.RespectHeadings {
		sections = c.splitByHeadings(content)
	} else {
		sections = []sectionData{{"", content}}
	}

	var chunks []Chunk
	chunkIndex := 0
	charOffset := 0

	for _, section := range sections {
		sectionChunks := c.chunkSection(section.content, section.title, chunkIndex, charOffset)
		chunks = append(chunks, sectionChunks...)
		chunkIndex += len(sectionChunks)
		charOffset += len(section.content)
	}

	// Add overlap
	if c.Config.ChunkOverlap > 0 {
		chunks = c.addOverlap(chunks, content)
	}

	// Add chunk markers
	if c.Config.AddChunkMarkers {
		for i := range chunks {
			chunks[i].Content = fmt.Sprintf("%s\n\n<!-- chunk_boundary: %d -->", chunks[i].Content, i)
		}
	}

	return chunks
}

type sectionData struct {
	title   string
	content string
}

func (c *Chunker) extractFrontmatter(markdown string) (string, string) {
	if strings.HasPrefix(markdown, "---") {
		idx := strings.Index(markdown[3:], "---")
		if idx > 0 {
			frontmatter := strings.TrimSpace(markdown[3 : idx+3])
			content := strings.TrimSpace(markdown[idx+6:])
			return content, frontmatter
		}
	}
	return markdown, ""
}

func (c *Chunker) splitByHeadings(content string) []sectionData {
	pattern := regexp.MustCompile(`(?m)^(#{1,6})\s+(.+)$`)
	matches := pattern.FindAllStringSubmatchIndex(content, -1)

	if len(matches) == 0 {
		return []sectionData{{"", content}}
	}

	var sections []sectionData
	lastEnd := 0
	lastTitle := ""

	for _, match := range matches {
		// Content before this heading
		if match[0] > lastEnd {
			sectionContent := strings.TrimSpace(content[lastEnd:match[0]])
			if sectionContent != "" {
				sections = append(sections, sectionData{lastTitle, sectionContent})
			}
		}
		lastTitle = strings.TrimSpace(content[match[4]:match[5]]) // Group 2
		lastEnd = match[0]
	}

	// Remaining content
	if lastEnd < len(content) {
		remaining := strings.TrimSpace(content[lastEnd:])
		if remaining != "" {
			sections = append(sections, sectionData{lastTitle, remaining})
		}
	}

	if len(sections) == 0 {
		sections = []sectionData{{"", content}}
	}

	return sections
}

func (c *Chunker) chunkSection(content, sectionTitle string, startIndex, charOffset int) []Chunk {
	var chunks []Chunk

	lines := strings.Split(content, "\n")
	currentChunk := ""
	currentStart := charOffset
	chunkIdx := startIndex

	i := 0
	for i < len(lines) {
		line := lines[i]

		// Check for protected blocks (tables, code blocks)
		if strings.HasPrefix(strings.TrimSpace(line), "```") {
			// Find end of code block
			blockLines := []string{line}
			i++
			for i < len(lines) && !strings.HasPrefix(strings.TrimSpace(lines[i]), "```") {
				blockLines = append(blockLines, lines[i])
				i++
			}
			if i < len(lines) {
				blockLines = append(blockLines, lines[i])
			}
			block := strings.Join(blockLines, "\n")

			if len(currentChunk)+len(block) > c.Config.ChunkSize && currentChunk != "" {
				chunks = append(chunks, c.createChunk(
					strings.TrimSpace(currentChunk),
					chunkIdx,
					currentStart,
					currentStart+len(currentChunk),
					sectionTitle,
				))
				chunkIdx++
				currentChunk = ""
				currentStart = charOffset + c.countChars(lines[:i])
			}
			currentChunk += block + "\n"
			i++
			continue
		}

		// Check for table
		if strings.HasPrefix(strings.TrimSpace(line), "|") && strings.HasSuffix(strings.TrimSpace(line), "|") {
			tableLines := []string{line}
			i++
			for i < len(lines) && strings.HasPrefix(strings.TrimSpace(lines[i]), "|") {
				tableLines = append(tableLines, lines[i])
				i++
			}
			block := strings.Join(tableLines, "\n")

			if len(currentChunk)+len(block) > c.Config.ChunkSize && currentChunk != "" {
				chunks = append(chunks, c.createChunk(
					strings.TrimSpace(currentChunk),
					chunkIdx,
					currentStart,
					currentStart+len(currentChunk),
					sectionTitle,
				))
				chunkIdx++
				currentChunk = ""
				currentStart = charOffset + c.countChars(lines[:i-len(tableLines)])
			}
			currentChunk += block + "\n"
			continue
		}

		// Regular line
		if len(currentChunk)+len(line) > c.Config.ChunkSize && currentChunk != "" {
			chunks = append(chunks, c.createChunk(
				strings.TrimSpace(currentChunk),
				chunkIdx,
				currentStart,
				currentStart+len(currentChunk),
				sectionTitle,
			))
			chunkIdx++
			currentChunk = ""
			currentStart = charOffset + c.countChars(lines[:i])
		}
		currentChunk += line + "\n"
		i++
	}

	if strings.TrimSpace(currentChunk) != "" {
		chunks = append(chunks, c.createChunk(
			strings.TrimSpace(currentChunk),
			chunkIdx,
			currentStart,
			currentStart+len(currentChunk),
			sectionTitle,
		))
	}

	return chunks
}

func (c *Chunker) createChunk(content string, index, startChar, endChar int, sectionTitle string) Chunk {
	hasTable := strings.Contains(content, "|") && strings.Contains(content, "---")
	hasImage := strings.Contains(content, "![") || strings.Contains(content, "[Image:")
	hasCode := strings.Contains(content, "```")

	// Determine content type
	contentType := "text"
	if hasTable {
		contentType = "table"
	} else if hasCode {
		contentType = "code"
	} else if hasImage {
		contentType = "image"
	}

	return Chunk{
		Content:   content,
		Index:     index,
		StartChar: startChar,
		EndChar:   endChar,
		Metadata: ChunkMetadata{
			SectionTitle: sectionTitle,
			HeadingPath:  c.extractHeadingPath(content),
			HasTable:     hasTable,
			HasImage:     hasImage,
			HasCode:      hasCode,
			ContentType:  contentType,
		},
	}
}

func (c *Chunker) extractHeadingPath(content string) []string {
	var headings []string
	for _, line := range strings.Split(content, "\n") {
		if strings.HasPrefix(line, "#") {
			text := strings.TrimLeft(line, "#")
			headings = append(headings, strings.TrimSpace(text))
		}
	}
	return headings
}

func (c *Chunker) addOverlap(chunks []Chunk, _ string) []Chunk {
	if len(chunks) <= 1 {
		return chunks
	}

	overlapSize := c.Config.ChunkOverlap

	for i := 1; i < len(chunks); i++ {
		prevChunk := chunks[i-1]
		content := prevChunk.Content

		// Get overlap from end of previous chunk
		overlapText := content
		if len(content) > overlapSize {
			overlapText = content[len(content)-overlapSize:]
		}

		// Find a good break point
		for _, breakStr := range []string{"\n\n", ". ", "\n"} {
			idx := strings.Index(overlapText, breakStr)
			if idx > 0 {
				overlapText = overlapText[idx+len(breakStr):]
				break
			}
		}

		if strings.TrimSpace(overlapText) != "" {
			chunks[i].Content = fmt.Sprintf("[...] %s\n\n%s", strings.TrimSpace(overlapText), chunks[i].Content)
		}
	}

	return chunks
}

func (c *Chunker) countChars(lines []string) int {
	total := 0
	for _, line := range lines {
		total += len(line) + 1 // +1 for newline
	}
	return total
}
